pkg/querier/timeseries_series_set.go

```diff
@@ -0,0 +1,99 @@
+package querier
+
+import (
+	"sort"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+// timeSeriesSeriesSet is a wrapper around a client.TimeSeries slice to implement to SeriesSet interface
+type timeSeriesSeriesSet struct {
+	ts []client.TimeSeries
+	i  int
+}
+
+func newTimeSeriesSeriesSet(series []client.TimeSeries) *timeSeriesSeriesSet {
+	sort.Sort(byTimeSeriesLabels(series))
+	return &timeSeriesSeriesSet{
+		ts: series,
+		i:  -1,
+	}
+}
+
+// Next implements SeriesSet interface
+func (t *timeSeriesSeriesSet) Next() bool { t.i++; return t.i < len(t.ts) }
+
+// At implements SeriesSet interface
+func (t *timeSeriesSeriesSet) At() storage.Series {
+	if t.i < 0 {
+		return nil
+	}
+	return &Timeseries{series: t.ts[t.i]}
+}
+
+// Err implements SeriesSet interface
+func (t *timeSeriesSeriesSet) Err() error { return nil }
+
+// Timeseries is a type wrapper that implements the storage.Series interface
+type Timeseries struct {
+	series client.TimeSeries
+}
+
+// TimeSeriesSeriesIterator is a wrapper around a client.TimeSeries to implement the SeriesIterator interface
+type TimeSeriesSeriesIterator struct {
+	ts *Timeseries
+	i  int
+}
+
+type byTimeSeriesLabels []client.TimeSeries
+
+func (b byTimeSeriesLabels) Len() int      { return len(b) }
+func (b byTimeSeriesLabels) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b byTimeSeriesLabels) Less(i, j int) bool {
+	return labels.Compare(client.FromLabelAdaptersToLabels(b[i].Labels), client.FromLabelAdaptersToLabels(b[j].Labels)) < 0
+}
+
+// Labels implements the storage.Series interface.
+// Conversion is safe because ingester sets these by calling client.FromLabelsToLabelAdapters which guarantees labels are sorted.
+func (t *Timeseries) Labels() labels.Labels {
+	return client.FromLabelAdaptersToLabels(t.series.Labels)
+}
+
+// Iterator implements the storage.Series interface
+func (t *Timeseries) Iterator() storage.SeriesIterator {
+	return &TimeSeriesSeriesIterator{
+		ts: t,
+		i:  -1,
+	}
+}
+
+// Seek implements SeriesIterator interface
+func (t *TimeSeriesSeriesIterator) Seek(s int64) bool {
+	idx := sort.Search(len(t.ts.series.Samples), func(i int) bool {
+		return t.ts.series.Samples[i].TimestampMs >= s
+	})
+
+	if idx < len(t.ts.series.Samples) {
+		t.i = idx
+		return true
+	}
+
+	return false
```

Comment:
To keep consistency with other iterator implementations, if a `Seek()` doesn't find any sample, `t.i` should be updated anyway.

Which means the function could be implemented as:
```
	t.i = sort.Search(len(t.ts.series.Samples), func(i int) bool {
		return t.ts.series.Samples[i].TimestampMs >= s
	})

        return t.i < len(t.ts.series.Samples)
```

_Please also add a test to cover this case._

Reply:
Actually, this implementation of `Seek` can go backwards, which is weird. `Seek` is only supposed to go forward. (On Prometheus master, this is now explicit: `Seek advances the iterator forward to the first sample with the timestamp equal or greater than t.`)

Reply:
I've updated seek to only advance, and added unit tests around Seek.